### ISSUE TITLE
[Patterns] Mixed multiple pattern and when is not fully supported in switch expression

### ISSUE THREAD

**Issue by NolwennD** [eclipse-jdt/eclipse.jdt.core#3109]
Since [JEP 456](https://openjdk.org/jeps/456) released in JDK 22 it’s possible to do this :
```java
switch (box) {
    case Box(RedBall _), Box(BlueBall _) -> processBox(box);
    case Box(GreenBall _)                -> stopProcessing();
    case Box(var _)                      -> pickAnotherBox();
}
```
But Eclipse detect some errors on this valid code
```java
package maybe;

import java.util.function.Function;
import java.util.function.Predicate;

public sealed interface Maybe<T>  {

  static <T> Maybe<T> of(T value) {
    return new Some<>(value);
  }

  static <T> Maybe<T> empty() {
    return new None<>();
  }

  default Maybe<T> filter(Predicate<T> predicate){
    return switch (this) {
      case Some(T value) when predicate.test(value) -> this; // unsupported line with 4 errors
      case Some(_), None<T> _ -> empty();
    };
  }
  
  default Maybe<T> filter2(Predicate<T> predicate){
    return switch (this) {
      case Some(_), None<T> _ -> empty(); //also unsupported with 1 error
    };
  }
  record Some<T>(T value) implements Maybe<T> {}
  record None<T>() implements Maybe<T>{}
}
```
### Eclipse
Version: 2024-09 (4.33.0)
Build id: 20240905-0614

**Comment by srikanth-sankaran** [comment]
I'll take a look. Thanks for the defect report.

**Comment by srikanth-sankaran** [comment]
Compiles fine on master/HEAD. Very likely a duplicate of #3003 - I'll verify, add a test case and close.

**Comment by srikanth-sankaran** [comment]
Yes, If I move back Scanner.java to the prior version to the fix for #3003, I get these errors:

```
Cannot mix pattern with other case labels
Invalid expression as statement
Syntax error, insert ":" to complete SwitchLabel
Syntax error, insert ":" to complete SwitchLabel
Syntax error on token "when", , expected
```
